consensus/spos/bls/v2/subroundBlock.go

```diff
@@ -0,0 +1,746 @@
+		return false, nil
+	}
+
+	prevHeader, prevHash := sr.getPrevHeaderAndHash()
+	if check.IfNil(prevHeader) {
+		return false, nil
+	}
+	if !bytes.Equal(header.GetPrevHash(), prevHash) {
+		return false, nil
+	}
+	if header.GetNonce() != prevHeader.GetNonce()+1 {
+		return false, nil
+	}
+	prevRandSeed := prevHeader.GetRandSeed()
+
+	return bytes.Equal(header.GetPrevRandSeed(), prevRandSeed), prevHeader
+}
+
+func (sr *subroundBlock) getLeaderForHeader(headerHandler data.HeaderHandler) ([]byte, error) {
+	nc := sr.NodesCoordinator()
+
+	prevBlockEpoch := uint32(0)
+	if sr.Blockchain().GetCurrentBlockHeader() != nil {
+		prevBlockEpoch = sr.Blockchain().GetCurrentBlockHeader().GetEpoch()
+	}
+	// TODO: remove this if first block in new epoch will be validated by epoch validators
+	// first block in epoch is validated by previous epoch validators
+	selectionEpoch := headerHandler.GetEpoch()
+	if selectionEpoch != prevBlockEpoch {
+		selectionEpoch = prevBlockEpoch
+	}
+	leader, _, err := nc.ComputeConsensusGroup(
+		headerHandler.GetPrevRandSeed(),
+		headerHandler.GetRound(),
+		headerHandler.GetShardID(),
+		selectionEpoch,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return leader.PubKey(), err
+}
+
+func (sr *subroundBlock) receivedBlockHeader(headerHandler data.HeaderHandler) {
+	if check.IfNil(headerHandler) {
+		return
+	}
+
+	log.Debug("subroundBlock.receivedBlockHeader", "nonce", headerHandler.GetNonce(), "round", headerHandler.GetRound())
+	if headerHandler.CheckFieldsForNil() != nil {
+		return
+	}
+
+	isHeaderForCurrentConsensus, prevHeader := sr.isHeaderForCurrentConsensus(headerHandler)
+	if !isHeaderForCurrentConsensus {
+		log.Debug("subroundBlock.receivedBlockHeader - header is not for current consensus")
+		return
+	}
+
+	isLeader := sr.IsSelfLeader()
+	if sr.ConsensusGroup() == nil || isLeader {
+		log.Debug("subroundBlock.receivedBlockHeader - consensus group is nil or is leader")
+		return
+	}
+
+	if sr.IsConsensusDataSet() {
+		log.Debug("subroundBlock.receivedBlockHeader - consensus data is set")
+		return
+	}
+
+	headerLeader, err := sr.getLeaderForHeader(headerHandler)
+	if err != nil {
+		log.Debug("subroundBlock.receivedBlockHeader - error getting leader for header", err.Error())
+		return
+	}
+
+	if !sr.IsNodeLeaderInCurrentRound(string(headerLeader)) {
+		sr.PeerHonestyHandler().ChangeScore(
+			string(headerLeader),
+			spos.GetConsensusTopicID(sr.ShardCoordinator()),
+			spos.LeaderPeerHonestyDecreaseFactor,
+		)
+
+		log.Debug("subroundBlock.receivedBlockHeader - leader is not the leader in current round")
+		return
+	}
+
+	if sr.IsHeaderAlreadyReceived() {
+		log.Debug("subroundBlock.receivedBlockHeader - header is already received")
+		return
+	}
+
+	if !sr.CanProcessReceivedHeader(string(headerLeader)) {
+		log.Debug("subroundBlock.receivedBlockHeader - can not process received header")
+		return
+	}
+
+	marshalledHeader, err := sr.Marshalizer().Marshal(headerHandler)
```

Comment:
so you need the hash only. there is a function which calculates that with one line of code.
core.CalculateHash

Reply:
changed